### PR TITLE
Empirical distribution

### PR DESCRIPTION
**Pull request by rnugent3**
This code is in draft status. I learned during testing that binWidth returns 0, even when I hard wire (in multiple locations) binWidth to return 1. @HenryGeorgist, would you take a look at the code to see where the issue might be? I've stared at this so long it turned up in my dreams.

### CODE REVIEW

File: statistics/empiricalDistribution.go [hunk 1]
```diff
@@ -128,14 +128,16 @@ func (e EmpiricalDistribution) CDF(value float64) float64 {
 }
 
 func (e EmpiricalDistribution) PDF(value float64) float64 {
```
Comment (HenryGeorgist) [inline statistics/empiricalDistribution.go:130]: try func(e *EmpericalDistribution) PDF(value float64) float64

**Comment by rnugent3** [comment]
All tests run

**Comment by rnugent3** [comment]
made revision to inline-histogram.go. I am not sure how to easily test if this code is not in master. Once in master, I can refine if needed.

File: data/inline-histogram.go [hunk 2]
```diff
@@ -178,6 +179,20 @@ func (ih *InlineHistogram) String() string {
 	}
 	return s
 }
+
+func (ih *InlineHistogram) HistogramToEmpiricalString() string {
```
Comment (HenryGeorgist) [inline data/inline-histogram.go:183]: it seems like you can produce an Empirical distribution from bins and bincounts - why not just create an exported function to access that protected memory in the struct?

**Review by HenryGeorgist — CHANGES_REQUESTED** [review]
It seems like the EmpiricalString Method is not necessary - lets create a func (ih inline-histogram) Bins() float[] and func (ih inline-histogram) BinCounts() float[] that will be faster and the conversion from float -> string -> float is a possible loss of information which is unnecessary in this case.

**Comment by HenryGeorgist** [comment]
> I am not sure how to easily test

I think we need to talk about how to test on a branch! Your tests pass on the branch for Empirical Distribution, why not add a test for inline histogram on your branch to test the behavior you added? I might be missing something with how you have set up your testing workflow though, so lets talk about it.

File: data/inline-histogram.go [hunk 3]
```diff
@@ -180,17 +181,20 @@ func (ih *InlineHistogram) String() string {
...
 
-func (ih *InlineHistogram) HistogramToEmpiricalString() string {
-	s := fmt.Sprintf("InlineHistogram:\nBinCount: %v\nObservations: %v\nMin: %f\nMax: %f\nMean: %f\n", len(ih.bins), ih.pm.sampleSize, ih.pm.min, ih.pm.max, ih.pm.GetMean())
-	s += "Bin Start"
-	for idx, _ := range ih.bins {
-		s += fmt.Sprintf("%f, ", ih.minValue+(ih.binWidth*float64(idx)))
+func (ih *InlineHistogram) Bins() []float64 {
+	binStarts := []float64{}
+	for i, _ := range ih.bins {
+		binStarts = append(binStarts, float64(i))
```
Comment (HenryGeorgist) [inline data/inline-histogram.go:187]: shouldnt this start with  min bin value and increment by bin width?

File: data/inline-histogram.go [hunk 4]
```diff
@@ -180,17 +181,20 @@ func (ih *InlineHistogram) String() string {
...
 	}
-	s += "\nBin Count:"
+	return binStarts
+}
+
+func (ih *InlineHistogram) BinCounts() []float64 {
+	binCounts := []float64{}
```
Comment (HenryGeorgist) [inline data/inline-histogram.go:193]: why not just return bins?

File: data/inline-histogram_test.go [hunk 5]
```diff
@@ -168,3 +168,31 @@ func TestAddObservationNormalDistConvergenceMostStrict(t *testing.T) {
 	fmt.Println(fmt.Sprintf("variance %f", ih.pm.GetSampleVariance()))
 	fmt.Println(fmt.Sprintf("sample size %d", ih.pm.GetSampleSize()))
 }
+
+func TestBins(t *testing.T) {
+	ih := Init(1.0, 0.0, 5.0)
```
Comment (HenryGeorgist) [inline data/inline-histogram_test.go:173]: bin width of 1 with a min value of zero should pass the test... try a min value of not zero and a width of not 1.

**Review by HenryGeorgist — CHANGES_REQUESTED** [review]

File: data/inline-histogram.go (hunk 3 above)
Comment (rnugent3) [reply data/inline-histogram.go:187]: Is the index of ih.bins not the same as the bin starts (and thereby the ih.minValue+ increments of ih.binWidth)?

File: data/inline-histogram.go (hunk 4 above)
Comment (rnugent3) [reply data/inline-histogram.go:193]: I think the answer depends on your response to the above comment. I might misunderstand what the index of ih.bins represents.

File: data/inline-histogram.go (hunk 3 above)
Comment (HenryGeorgist) [reply data/inline-histogram.go:187]: the indexes are the same but to calculate bin starts you have to offset by the min value and multiply by bin width.

File: data/inline-histogram.go (hunk 4 above)
Comment (HenryGeorgist) [reply data/inline-histogram.go:193]: it is the count of values at a range the size of bin width on the real number line. to calculate the position use the min value + index)*binwidth  (assuming index starts at zero).

File: data/inline-histogram.go (hunk 3 above)
Comment (HenryGeorgist) [reply data/inline-histogram.go:187]: line 187 should contain something like ih.minValue + (float64(i)*ih.binWidth)

Comment (HenryGeorgist) [reply data/inline-histogram.go:187]: also, append is probably slow, you know the count based on the length of bins, so you can preallocate the array to the right size.

File: data/inline-histogram_test.go (hunk 5 above)
Comment (rnugent3) [reply data/inline-histogram_test.go:173]: Done. I'll push after I fix the above issues.

**Review by HenryGeorgist — APPROVED** [review]

**Pull request merged**